File: designpattern/python/CreationalPattern/AbstarctFactory/AbstractFactory.py
# Python Design Pattern : AbstractFactory
# Author:
# 2025

import sys

# 제품에 인터페이스
class ProductA:
    def getName(self):
        pass

# 제품 스펙 구현체
class ConcreateProductAX(ProductA):
    def getName(self):
        return "A-X"

class ConcreateProductAY(ProductA):
    def getName(self):
        return "A-Y"


class ProductB:
    def getName(self):
        pass

class ConcreateProductBX(ProductB):
    def getName(self):
        return "B-X"

class ConcreateProductBY(ProductB):
    def getName(self):
        return "B-Y"


# 팩토리 정의 인터페이스, 생성 책임
class AbstractFactory:
    def createProductA(self):
        pass
    
    def createProductB(self):
        pass

# 팩토리 구현체
class ConcreateFactoryX(AbstractFactory):
    def createProductA(self):
        return ConcreateProductAX()
    
    def createProductB(self):
        return ConcreateProductBX()
    
class ConcreateFacotryY(AbstractFactory):
    def createProductA(self):
        return ConcreateProductAY()
    
    def createProductB(self):
        return ConcreateProductBY()
    

if __name__=='__main__':
    factoryX = ConcreateFactoryX()
    factoryY = ConcreateFacotryY()

    p1 = factoryX.createProductA()
    print ("Product: " + p1.getName())
    
    p2 = factoryY.createProductA()
    print ("Product: " + p2.getName())
    

File: designpattern/python/CreationalPattern/AbstarctFactory/AbstractFactory.cpp
#include <iostream>

class ProductA
{
public:
    virtual ~ProductA() {}
    
    virtual const char* getName() = 0;
};

class ConcreateProductAX : public ProductA
{
public:
  ~ConcreateProductAX() {}
  
  const char* getName()
  {
    return "A-X";
  }
};

class ConcreateProductAY : public ProductA
{
public:
  ~ConcreateProductAY() {}
  
  const char* getName()
  {
    return "A-Y";
  }
};



class ProductB
{
public:
  virtual ~ProductB() {}
  
  virtual const char* getName() = 0;
};

class ConcreateProductBX : public ProductB
{
public:
  ~ConcreateProductBX () {}
  
  const char* getName()
  {
    return "B-X";
  }
};

class ConcreateProductBY : public ProductB
{
public:
  ~ConcreateProductBY () {}
  
  const char* getName()
  {
    return "B-Y";
  }
};

class AbstractFactory
{
public:
  virtual ~AbstractFactory() {}
  
  virtual ProductA *createProductA() = 0;
  virtual ProductB *createProductB() = 0;
};

class ConcreateFactoryX : public AbstractFactory
{
public:
  ~ConcreateFactoryX() {}
  
  ProductA *createProductA()
  {
    return new ConcreateProductAX();
  }

  ProductB *createProductB()
  {
    return new ConcreateProductBX();
  }
};

class ConcreateFactoryY : public AbstractFactory
{
public:
  ~ConcreateFactoryY() {}
  
  ProductA *createProductA()
  {
    return new ConcreateProductAY();
  }
  
  ProductB *createProductB()
  {
    return new ConcreateProductBY();
  }
};

int main()
{
  ConcreateFactoryX *factoryX = new ConcreateFactoryX();
  ConcreateFactoryY *factoryY = new ConcreateFactoryY();

  ProductA *p1 = factoryX -> createProductA();
  std::cout << "Product: " << p1->getName() << std::endl;

  ProductB *p2 = factoryX -> createProductB();
  std::cout << "Product: " << p2->getName() << std::endl;

  delete p1;
  delete p2;

  delete factoryX;
  delete factoryY;

  return 0;
}